### PR TITLE
Implements ColumnPruning optimization rule

### PR DESCRIPTION
**Pull request by UranusSeven**
## What do these changes do?

Implements ColumnPruning rule to optimize tilebale graph.

## Related issue number

Fixes #xxxx

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/optimization/logical/common/head.py [hunk 1]
```diff
@@ -14,33 +14,34 @@
 
 from typing import List
 
+from ..core import (
```
Comment (aresnow1) [inline mars/optimization/logical/common/head.py:17]: Keep the import order.

File: mars/optimization/logical/tileable/__init__.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from .column_pruning import ColumnPruningRule
```
Comment (aresnow1) [inline mars/optimization/logical/tileable/__init__.py:15]: Order of imports is incorrect.

**Review by aresnow1 — COMMENTED** [review]
Order of import statements is incorrect at some places.

File: mars/optimization/logical/core.py [hunk 3]
```diff
@@ -175,37 +160,88 @@ def _remove_collapsable_predecessors(self, node: EntityType):
...
+            op = entity.op
+            if op in visited:
+                continue
+            visited.add(op)
+
+            if entity not in self._graph:  # pragma: no cover
+                # maybe removed during optimization
+                continue
+            op_types = self._rule_type_to_op_types[type(self)]
+            if isinstance(op, tuple(op_types)) and self.match_operand(op):
```
[the remark this answers is not in the document]
Comment (UranusSeven) [reply mars/optimization/logical/core.py:194]: Yes, passing the right operand to a rule was handled by the framework. So`OperandBasedOptimizationRule` needs to do the same thing.

File: mars/optimization/logical/core.py [hunk 4]
```diff
@@ -175,37 +160,88 @@ def _remove_collapsable_predecessors(self, node: EntityType):
...
+    """
 
-    @classmethod
-    def register_rule(
-        cls, operand_types: List[Type[OperandType]], rule: Type[OptimizationRule]
+    _rule_type_to_op_types: Dict[
+        Type[OptimizationRule], Set[Type[OperandType]]
+    ] = defaultdict(set)
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (UranusSeven) [reply mars/optimization/logical/core.py]: Removed.

File: mars/core/operand/base.py [hunk 5]
```diff
@@ -232,7 +232,7 @@ def _set_inputs(self, inputs):
         setattr(self, "_inputs", inputs)
 
     @property
-    def inputs(self) -> List[Union[Chunk, Tileable]]:
+    def inputs(self) -> List[Union[ENTITY_TYPE]]:
```
[the remark this answers is not in the document]
Comment (UranusSeven) [reply mars/core/operand/base.py]: Fixed.

File: mars/optimization/logical/common/head.py (hunk 1 above)
Comment (UranusSeven) [reply mars/optimization/logical/common/head.py]: Fixed.

**Review by aresnow1 — APPROVED** [review]
LGTM

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Pull request merged**